pkg/code-generator/schemagen/generator.go

```diff
@@ -85,60 +85,62 @@ func GenerateOpenApiValidationSchemas(project *model.Project, options *Validatio
 		return err
 	}
 
-	// For each matching CRD, apply the JSON schema to that CRD
+	// For each version of each matching CRD, apply the JSON schema to that CRD
 	// Use Group.Version.Kind to match CRDs and Schemas
 	crdWriter := NewCrdWriter(options.CrdDirectory)
 	for _, crd := range crds {
-		crdGVK := schema.GroupVersionKind{
-			Group:   crd.Spec.Group,
-			Version: crd.Spec.Version,
-			Kind:    crd.Spec.Names.Kind,
-		}
+		for i := 0; i < len(crd.Spec.Versions); i++ {
+			crdGVK := schema.GroupVersionKind{
+				Group:   crd.Spec.Group,
+				Version: crd.Spec.Versions[i].Name,
+				Kind:    crd.Spec.Names.Kind,
+			}
 
-		specJsonSchema, ok := jsonSchemasByGVK[crdGVK]
-		if !ok {
-			continue
-		}
+			specJsonSchema, ok := jsonSchemasByGVK[crdGVK]
+			if !ok {
+				continue
+			}
 
-		// prevent k8s from validating metadata field
-		removeProtoMetadataValidation(specJsonSchema)
+			// prevent k8s from validating metadata field
+			removeProtoMetadataValidation(specJsonSchema)
 
-		if err := validateStructural(crdGVK, specJsonSchema); err != nil {
-			return err
-		}
+			if err := validateStructural(crdGVK, specJsonSchema); err != nil {
+				return err
+			}
 
-		validationSchema := &apiextv1beta1.CustomResourceValidation{
-			OpenAPIV3Schema: &apiextv1beta1.JSONSchemaProps{
-				Type:       "object",
-				Properties: map[string]apiextv1beta1.JSONSchemaProps{},
-			},
-		}
+			validationSchema := &apiextv1.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextv1.JSONSchemaProps{
+					Type:       "object",
+					Properties: map[string]apiextv1.JSONSchemaProps{},
+				},
+			}
 
-		// Either use the status defined on the spec, or a generic status
-		statusSchema := specJsonSchema.Properties["status"]
-		if statusSchema.Type == "" {
-			statusSchema = apiextv1beta1.JSONSchemaProps{
-				Type:                   "object",
-				XPreserveUnknownFields: pointer.BoolPtr(true),
+			// Either use the status defined on the spec, or a generic status
+			statusSchema := specJsonSchema.Properties["status"]
+			if statusSchema.Type == "" {
+				statusSchema = apiextv1.JSONSchemaProps{
+					Type:                   "object",
+					XPreserveUnknownFields: pointer.BoolPtr(true),
+				}
 			}
-		}
 
-		validationSchema.OpenAPIV3Schema.Properties["spec"] = *specJsonSchema
-		validationSchema.OpenAPIV3Schema.Properties["status"] = statusSchema
+			validationSchema.OpenAPIV3Schema.Properties["spec"] = *specJsonSchema
+			validationSchema.OpenAPIV3Schema.Properties["status"] = statusSchema
 
-		if err = crdWriter.ApplyValidationSchemaToCRD(crd, validationSchema); err != nil {
-			return err
+			if err = crdWriter.ApplyValidationSchemaToCRD(crd, validationSchema); err != nil {
```

Comment:
ApplyValidationSchemaToCRD was probably a poorly named function by me in the first place, but this will write the validation schema for a particualr CRD to a file. Right now we iterate over each version (even though we only support 1) and always write that to the 1st validation schema. Therefore, if we supported multiple versions, they would each be overwritten. 

Can we either:
(A) explicitly only support 1 and throw an error if there are more than 1
or 
(B) change how we call the ApplyValidationSchemaToCRD to support validation schemas per version